fw/ss_skb.c

```diff
@@ -1008,6 +1008,94 @@ ss_skb_chop_head_tail(struct sk_buff *skb_head, struct sk_buff *skb,
 	return 0;
 }
 
+/**
+ * Chop @head and @tail bytes at head and end of the message (skb circular
+ * list), iterating over the list if nessessary.
+ * @return 0 on success and -(error code) on error.
+ *
+ * Message length should be greater than the sum of @head and @trail,
+ * otherwise an error is retuned.
+ */
+int
+ss_skb_list_chop_head_tail(struct sk_buff **skb_list_head, 
+			   size_t head, size_t trail)
+{
+
+	struct sk_buff *skb, *skb_hd;
+	size_t sum;
+	int R;
+	skb = *skb_list_head;
+	if (likely(skb->next == skb))
+		goto single_buff;
+	if (likely(head))
+		goto head_and_probably_trail;
+	if (likely(trail))
+		goto trail_only;
+	/* nothing to chop */
+	return 0;
+
+single_buff:
+	/* There is the only 1 buffer in the list */
+	sum = head + trail;
+	if (WARN_ON(skb->len <= sum))
+		return -EBADMSG;
+	if (sum == 0)
+		/* Nothing to chop */
+		/* This check is mostly for jumps from branches below */
+		return 0;
+	return ss_skb_chop_head_tail(skb, skb, head, trail);
+
+head_and_probably_trail:
+	do {
+		if (likely(skb->len > head)) {
+			R = ss_skb_chop_head_tail(skb, skb, head, 0);
+			if (unlikely(R < 0))
+				return R;
+			if (unlikely(trail)) {
+				head = 0;
+				goto trail_only;
+			}
+			return 0;
+		}
+		head -= skb->len;
+		/* We do not use ss_skb_unlink() here to prevent it
+	 	* to remove the last skb in the list and to skip
+	 	* some actions using our apriori knowledge
+	 	*/
```

Comment:
I stuck again on the code, why we don't use `ss_skb_unlink()` in particular. I.e. why it's bad to have empty list after its call. In https://github.com/tempesta-tech/tempesta/pull/1534#discussion_r765725584 I commented that `ss_skb_unlink()` just can silently make the skb list empty and we'll need to check this along with other extra checks in `ss_skb_unlink()`. Please improve the comment to make it more clear. But it's not necessary to duplicate the comment in single small function twice.

Reply:
Seen. Thinknig...

Reply:
Probably we can use skb_unlink(), at least now, when we guarantee that the list does not contain the only one skb/

Reply:
Yeah, need to handle the question carefully - the code is crucial and we need to accurately rework it and properly document with comments

Reply:
> Please improve the comment to make it more clear.

I have tried